File: cpp/min_diff_bst_530.cpp
#include <iostream>
#include <climits>
#include <algorithm>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int min_diff = INT_MAX;
    int previous = -1;
    int getMinimumDifference(TreeNode *root) {
        if (root->left != NULL) getMinimumDifference(root->left);
        if (previous >= 0) min_diff = min(min_diff, root->val - previous);
        previous = root->val;
        if (root->right != NULL) getMinimumDifference(root->right);
        return min_diff;
    }
};

int main() {
    TreeNode *root = new TreeNode(1);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(2);
    cout << (new Solution())->getMinimumDifference(root) << endl;
    return 0;
}

File: python/min_diff_bst_530.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.previous = -1
        self.ans = float('inf')
        self.min_diff(root)
        return self.ans
    
    def min_diff(self, root):
        if root.left:
            self.min_diff(root.left)
        if self.previous >= 0:
            self.ans = min(self.ans, root.val - self.previous)
        self.previous = root.val
        if root.right:
            self.min_diff(root.right)